### DISCUSSION TITLE
How to send additional data in error response?

### DISCUSSION THREAD

**Discussion by thobson** [swaggest/rest#152] [Q&A]
I'm probably missing something here but in my case the error responses don't match the format described by swagger.

For example:

```go
u := usecase.NewInteractor(func(ctx context.Context, input TenancyInput, output *User) error {
  return status.Wrap(errors.New("record not found"), status.NotFound)
}
u.SetExpectedErrors(status.NotFound)
```

Generates swagger like this:

```json
{
  "responses": {
    "404": {
     "description": "Not Found",
     "content": {
      "application/json": {
       "schema": {
        "$ref": "#/components/schemas/RestErrResponse"
       }
      }
     }
    }
   },
   "RestErrResponse": {
    "type": "object",
    "properties": {
     "code": {
      "type": "integer",
      "description": "Application-specific error code."
     },
     "context": {
      "type": "object",
      "additionalProperties": {},
      "description": "Application context."
     },
     "error": {
      "type": "string",
      "description": "Error message."
     },
     "status": {
      "type": "string",
      "description": "Status text."
     }
    }
   }
}
```

i.e status, error, code, context

But the server actually sends

```json
{
  "status": "NOT_FOUND",
  "error": "not found: record not found"
}
``` 

Is there anyway I can also send a code and additional context in the response?

Thanks

**Comment by vearutop** [comment]
Context of `error` is extracted via [`rest.ErrWithFields`](https://pkg.go.dev/github.com/swaggest/rest#ErrWithFields).

```go
type ErrWithFields interface {
	error
	Fields() map[string]interface{}
}
```

You can either implement your own structured error or use [`ctxd.NewError`](https://pkg.go.dev/github.com/bool64/ctxd#NewError) (external dependency).

**Comment by mtiller** [comment]
Am I correct in understanding that all errors generated by this framework will ultimately follow the `ErrResponse` schema?  I see that there is a way to specify the schema associated with a "successful result".  But is there no way to specify any other error schema besides `ErrResponse`?  For example, support for [RFC7807](https://www.rfc-editor.org/rfc/rfc7807).  It seems quite reasonable that an API developer would want to conform to RFC7807 but I don't see any way to do that here.

**Reply by vearutop** [reply]
`ErrResponse` is default structure for unsuccessful results, however any kind of structure can be used with a bit of instrumentation.

Please check an [example](https://github.com/swaggest/rest/blob/v0.2.45/_examples/advanced-generic/router.go#L92-L122), it sets up `(*Handler).MakeErrResp` to process use case errors.

Error content type can be customized with [`response.DefaultErrorResponseContentType`](`https://pkg.go.dev/github.com/swaggest/rest@v0.2.45/response#pkg-variables`).
```go
	response.DefaultErrorResponseContentType = "application/problem+json"
	s := web.DefaultService()
......
	s.Wrap(
		func(handler http.Handler) http.Handler {
			var h *nethttp.Handler
			if nethttp.HandlerAs(handler, &h) {
				h.MakeErrResp = func(ctx context.Context, err error) (int, interface{}) {
					httpStatus, rfc7807Err := makeRFC7807Error(err) // Your custom error factory.
					
					return httpStatus, rfc7807Err
				}
			}

			return handler
		},
	)
.....
// Mount use case handlers.
``` 


If you land at a convenient and useful implementation of a RFC7807 error, please feel free to submit a PR with such a middleware, it may be helpful to other folks.
